email/provider_test.go

```diff
@@ -57,6 +60,59 @@ func testAccCheckEmailExists(n string) resource.TestCheckFunc {
 		return nil
 	}
 }
+func mockSendMail(addr string, a smtp.Auth, from string, to []string, msg []byte) error {
+	return errors.New("421 Service not available")
+}
+
+// function to test non-421 error
+func mockReturnNon421(addr string, a smtp.Auth, from string, to []string, msg []byte) error {
+	return errors.New("500 Internal Server Error")
+}
+
+// email to return no error
+func mockSendMailSuccess(addr string, a smtp.Auth, from string, to []string, msg []byte) error {
+	return nil
+}
+
+func TestRetryWorkflow(t *testing.T) {
+	// define max retries
+	maxRetries := 5
+	// write the tests
+	tests := []struct {
+		name         string
+		sendMailFunc SendMailFunc
+		expectedErr  error
+	}{
+		{
+			name:         "Retry with 421 error",
+			sendMailFunc: mockSendMail,
+			expectedErr:  errors.New("421 Service not available"),
```

Comment:
Here we want to test that the retry is working, i.e. mockSendMail is being called more than once, right? Otherwise I don't seem to see anywhere that tests this behaviour.

Reply:
Deferred to another ticket